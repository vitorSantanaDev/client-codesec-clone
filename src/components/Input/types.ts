/* eslint-disable @typescript-eslint/no-explicit-any */
export type IViewProps = {
  id?: string
  type: string
  name: string
  value?: string
  valueIsValid?: boolean
  placeholder?: string
  errorMessage?: string
  handleChange: (event?: any) => void
}
