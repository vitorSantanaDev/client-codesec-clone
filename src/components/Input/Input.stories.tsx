import { ComponentStory, ComponentMeta } from '@storybook/react'
import Input from '.'

export default {
  title: 'Input',
  component: Input,
  args: {
    id: 'email',
    value: '',
    type: 'email',
    name: 'email',
    valueIsValid: false,
    placeholder: 'Digite seu email',
    handleChange: ({ target }) => console.log(target)
  }
} as ComponentMeta<typeof Input>

export const Default: ComponentStory<typeof Input> = (args) => {
  return <Input {...args} />
}

export const ErrorMessage: ComponentStory<typeof Input> = (args) => {
  return <Input {...args} />
}

ErrorMessage.args = {
  errorMessage: 'Aqui vai a mensagem de erro!',
  valueIsValid: false
}

export const ValueIsValid: ComponentStory<typeof Input> = (args) => {
  return <Input {...args} />
}

ValueIsValid.args = {
  valueIsValid: true,
  type: 'text',
  name: 'name'
}
