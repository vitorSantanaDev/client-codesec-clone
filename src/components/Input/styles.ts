import styled, { css } from 'styled-components'

type InputProps = {
  errorMessage?: string
  valueIsValid?: boolean
}

export const Wrapper = styled.div`
  ${() => css`
    position: relative;

    display: flex;
    justify-content: space-between;
  `}
`
export const InputElement = styled.input<InputProps>`
  ${({ theme, errorMessage, valueIsValid }) => css`
    width: 100%;
    display: block;
    padding: calc(${theme.spacings.small} - 0.4rem) 0
      calc(${theme.spacings.small} - 0.4rem) ${theme.spacings.xlarge};

    border: none;
    outline: none;
    border-radius: 4px;

    font-size: ${theme.font.sizes.small};
    color: ${theme.colors.gray7};
    background-color: ${theme.colors.blackLight};

    ${errorMessage
      ? `
      border: 1px solid ${theme.colors.redError};
    `
      : ''}
    ${valueIsValid && !errorMessage
      ? `
      border: 1px solid ${theme.colors.greenSucsses};
    `
      : ''}

      &:focus {
      background-color: ${theme.colors.black};
    }
  `}
`

export const InputStateWrapper = styled.div`
  ${() => css``}
`

export const IconState = styled.span<InputProps>`
  ${({ errorMessage, valueIsValid }) => css`
    > svg {
      position: absolute;
      ${errorMessage &&
      `
        top: 35px;
        right: 10px;
        `}
      ${valueIsValid &&
      `
        top: 32px;
        right: 12px;
        `}
    }
  `}
`

export const InputIcon = styled.span<InputProps>`
  ${({ errorMessage, valueIsValid }) => css`
    > svg {
      position: absolute;
      top: 32px;
      left: 10px;

      ${valueIsValid &&
      `
       left: 11px;
      `}
      ${errorMessage &&
      `
      top: 35px;
      `}
    }
  `}
`

export const ErrorMessage = styled.span`
  ${({ theme }) => css`
    display: inline-block;
    margin-bottom: calc(${theme.spacings.small} - 1rem);

    font-weight: 400;
    font-size: calc(${theme.font.sizes.small} - 0.2rem);
    color: ${theme.colors.redError};
  `}
`
