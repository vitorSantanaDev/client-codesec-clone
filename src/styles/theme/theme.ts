export const theme = {
  colors: {
    redPrimary: '#E83752',
    redLight: '#FA5770',
    redDark: '#D4203B',
    redError: '#EB3D3D',
    redGradient: 'linear-gradient(243.46deg, #FA5770 16.66%, #E83752 83.34%)',
    redGradientDark:
      'linear-gradient(225.05deg, #FA5770 0%, rgba(232, 55, 82, 0) 100%)',
    redGradientLight:
      'linear-gradient(225.05deg, rgba(250, 87, 112, 0.1) 0%, rgba(232, 55, 82, 0) 50%)',
    greenSucsses: '#32D957',
    white: '#FFFFFF',
    whiteLight: '#F5F0F1',
    whiteDark: '#EDE4E5',
    black: '#121111',
    blackLight: '#1C1B1B',
    blackDark: '#0A0909',
    gray1: '#141414',
    gray2: '#2C2C2C',
    gray3: '#444444',
    gray4: '#5C5C5C',
    gray5: '#747474',
    gray6: '#8C8C8C',
    gray7: '#A3A3A3'
  },
  font: {
    family: {
      default: "'Raleway', sans-serif"
    },
    sizes: {
      small: '1.6rem',
      medium: '2.4rem',
      large: '3.2rem',
      xlarge: '4.0rem',
      xxlarge: '4.8rem',
      huge: '5.6rem',
      xhuge: '6.4rem',
      xxhuge: '8rem'
    }
  },
  media: {
    lteMedium: '(max-width: 768px)',
    lteLarge: '(max-width: 820px)',
    ltexLarge: '(max-width: 912px)',
    lteSmall: '(max-width: 420px)',
    lteXSmall: '(max-width: 280px)'
  },
  spacings: {
    small: '1.6rem',
    medium: '2.4rem',
    large: '3.2rem',
    xlarge: '4.0rem',
    xxlarge: '4.8rem',
    huge: '5.6rem',
    xhuge: '6.4rem',
    xxhuge: '8rem'
  }
}
