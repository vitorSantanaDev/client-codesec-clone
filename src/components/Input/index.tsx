/* eslint-disable @typescript-eslint/no-explicit-any */
import { AtOutline } from '@styled-icons/evaicons-outline/AtOutline'
import { CloseOutline } from '@styled-icons/evaicons-outline/CloseOutline'
import { PersonOutline } from '@styled-icons/evaicons-outline/PersonOutline'
import { CheckmarkOutline } from '@styled-icons/evaicons-outline/CheckmarkOutline'

import { theme } from 'styles/theme/theme'
import * as S from './styles'

type InputProps = {
  id?: string
  type: string
  name: string
  value?: string
  valueIsValid?: boolean
  placeholder?: string
  errorMessage?: string
  handleChange: (event?: any) => void
}

const Input = ({
  id,
  type,
  name,
  value,
  placeholder,
  valueIsValid = false,
  errorMessage = '',
  handleChange
}: InputProps) => {
  const colorRedPrimary = theme.colors.redPrimary

  const inputIsValidOrNotValid =
    valueIsValid && !errorMessage ? (
      <CheckmarkOutline size={24} color={theme.colors.greenSucsses} />
    ) : errorMessage ? (
      <CloseOutline size={24} color={theme.colors.redError} />
    ) : null

  const inputIcon =
    name === 'email' ? (
      <AtOutline size={24} color={colorRedPrimary} />
    ) : name === 'name' ? (
      <PersonOutline size={24} color={colorRedPrimary} />
    ) : null

  const stateInputIcons = {
    errorMessage,
    valueIsValid
  }

  return (
    <>
      <S.Wrapper>
        <S.InputIcon {...stateInputIcons}>{inputIcon}</S.InputIcon>
        <S.InputStateWrapper>
          <S.IconState {...stateInputIcons}>
            {inputIsValidOrNotValid}
          </S.IconState>
          <S.ErrorMessage>{errorMessage}</S.ErrorMessage>
        </S.InputStateWrapper>
      </S.Wrapper>
      <S.InputElement
        type={type}
        name={name}
        value={value}
        id={id || name}
        placeholder={placeholder}
        onChange={handleChange}
        valueIsValid={valueIsValid}
        errorMessage={errorMessage || ''}
      />
    </>
  )
}

export default Input
