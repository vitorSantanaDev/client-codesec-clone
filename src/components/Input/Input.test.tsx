/* eslint-disable @typescript-eslint/ban-ts-comment */
import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderTheme } from 'styles/theme/render-theme'
import { theme } from 'styles/theme/theme'

import Input from '.'

const mockProps = {
  type: 'text',
  id: 'name',
  name: 'name',
  placeholder: 'Name',
  handleChange: jest.fn()
}

describe('<Input />', () => {
  it('should render the input with default styles', () => {
    const { container } = renderTheme(<Input {...mockProps} />)
    const input: HTMLInputElement = screen.getByPlaceholderText(/name/i)

    expect(input).toBeInTheDocument()
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render the input with error styles when error message exist', () => {
    const { container } = renderTheme(
      <Input {...mockProps} errorMessage="Error MEssage!" />
    )
    const input: HTMLInputElement = screen.getByPlaceholderText(/name/i)

    expect(input).toBeInTheDocument()
    expect(input).toHaveStyle({
      border: `1px solid ${theme.colors.redError}`
    })
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render input with styles when input value is valid', () => {
    const { container } = renderTheme(<Input {...mockProps} valueIsValid />)
    const input: HTMLInputElement = screen.getByPlaceholderText(/name/i)

    expect(input).toBeInTheDocument()
    expect(input).toHaveStyle({
      border: `1px solid ${theme.colors.greenSucsses}`
    })
    expect(container.firstChild).toMatchSnapshot()
  })
  it('should call handleChange function when typing in input', async () => {
    renderTheme(<Input {...mockProps} />)
    const input: HTMLInputElement = screen.getByPlaceholderText('Name')
    const newValue = 'New value'

    expect(input).toBeInTheDocument()
    await userEvent.type(input, newValue)
    expect(mockProps.handleChange).toBeCalledTimes(9)
    expect(input.value).toBe(newValue)
  })
})
