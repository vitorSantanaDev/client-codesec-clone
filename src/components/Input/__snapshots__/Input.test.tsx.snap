// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> should render input with styles when input value is valid 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 > svg {
  position: absolute;
  top: 32px;
  right: 12px;
}

.c1 > svg {
  position: absolute;
  top: 32px;
  left: 10px;
  left: 11px;
}

.c4 {
  display: inline-block;
  margin-bottom: calc(1.6rem - 1rem);
  font-weight: 400;
  font-size: calc(1.6rem - 0.2rem);
  color: #EB3D3D;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="c2"
      color="#E83752"
      fill="currentColor"
      focusable="false"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        data-name="Layer 2"
      >
        <g
          data-name="person"
        >
          <path
            d="M12 11a4 4 0 10-4-4 4 4 0 004 4zm0-6a2 2 0 11-2 2 2 2 0 012-2zM12 13a7 7 0 00-7 7 1 1 0 002 0 5 5 0 0110 0 1 1 0 002 0 7 7 0 00-7-7z"
          />
        </g>
      </g>
    </svg>
  </span>
  <div
    class=""
  >
    <span
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c2"
        color="#32D957"
        fill="currentColor"
        focusable="false"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Layer 2"
        >
          <path
            d="M9.86 18a1 1 0 01-.73-.32l-4.86-5.17a1 1 0 111.46-1.37l4.12 4.39 8.41-9.2a1 1 0 111.48 1.34l-9.14 10a1 1 0 01-.73.33z"
            data-name="checkmark"
          />
        </g>
      </svg>
    </span>
    <span
      class="c4"
    />
  </div>
</div>
`;

exports[`<Input /> should render the input with default styles 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 > svg {
  position: absolute;
}

.c1 > svg {
  position: absolute;
  top: 32px;
  left: 10px;
}

.c4 {
  display: inline-block;
  margin-bottom: calc(1.6rem - 1rem);
  font-weight: 400;
  font-size: calc(1.6rem - 0.2rem);
  color: #EB3D3D;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="c2"
      color="#E83752"
      fill="currentColor"
      focusable="false"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        data-name="Layer 2"
      >
        <g
          data-name="person"
        >
          <path
            d="M12 11a4 4 0 10-4-4 4 4 0 004 4zm0-6a2 2 0 11-2 2 2 2 0 012-2zM12 13a7 7 0 00-7 7 1 1 0 002 0 5 5 0 0110 0 1 1 0 002 0 7 7 0 00-7-7z"
          />
        </g>
      </g>
    </svg>
  </span>
  <div
    class=""
  >
    <span
      class="c3"
    />
    <span
      class="c4"
    />
  </div>
</div>
`;

exports[`<Input /> should render the input with error styles when error message exist 1`] = `
.c2 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c0 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c3 > svg {
  position: absolute;
  top: 35px;
  right: 10px;
}

.c1 > svg {
  position: absolute;
  top: 32px;
  left: 10px;
  top: 35px;
}

.c4 {
  display: inline-block;
  margin-bottom: calc(1.6rem - 1rem);
  font-weight: 400;
  font-size: calc(1.6rem - 0.2rem);
  color: #EB3D3D;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    <svg
      aria-hidden="true"
      class="c2"
      color="#E83752"
      fill="currentColor"
      focusable="false"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        data-name="Layer 2"
      >
        <g
          data-name="person"
        >
          <path
            d="M12 11a4 4 0 10-4-4 4 4 0 004 4zm0-6a2 2 0 11-2 2 2 2 0 012-2zM12 13a7 7 0 00-7 7 1 1 0 002 0 5 5 0 0110 0 1 1 0 002 0 7 7 0 00-7-7z"
          />
        </g>
      </g>
    </svg>
  </span>
  <div
    class=""
  >
    <span
      class="c3"
    >
      <svg
        aria-hidden="true"
        class="c2"
        color="#EB3D3D"
        fill="currentColor"
        focusable="false"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          data-name="Layer 2"
        >
          <path
            d="M13.41 12l4.3-4.29a1 1 0 10-1.42-1.42L12 10.59l-4.29-4.3a1 1 0 00-1.42 1.42l4.3 4.29-4.3 4.29a1 1 0 000 1.42 1 1 0 001.42 0l4.29-4.3 4.29 4.3a1 1 0 001.42 0 1 1 0 000-1.42z"
            data-name="close"
          />
        </g>
      </svg>
    </span>
    <span
      class="c4"
    >
      Error MEssage!
    </span>
  </div>
</div>
`;
